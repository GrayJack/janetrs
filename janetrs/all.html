<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-47e7ab555ef2818a.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="janetrs" data-themes="" data-resource-suffix="" data-rustdoc-version="1.74.0-nightly (030e4d382 2023-09-10)" data-channel="nightly" data-search-js="search-5d3eaacf19ebf04f.js" data-settings-js="settings-74424d7eec62a23e.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-f194925aa375ae96.css" data-theme-dark-css="dark-1dd4d1ce031e15de.css" data-theme-ayu-css="ayu-49e58d069f567085.css" ><script src="../static.files/storage-db41da1a38ea3cb8.js"></script><script defer src="../static.files/main-8d035c8cea6edbc4.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-f194925aa375ae96.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1dd4d1ce031e15de.css"><link rel="stylesheet" href="../static.files/noscript-cffde32267a19fd6.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate janetrs</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Aliases</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="struct.Janet.html">Janet</a></li><li><a href="struct.JanetAbstract.html">JanetAbstract</a></li><li><a href="struct.JanetArray.html">JanetArray</a></li><li><a href="struct.JanetBuffer.html">JanetBuffer</a></li><li><a href="struct.JanetConversionError.html">JanetConversionError</a></li><li><a href="struct.JanetFiber.html">JanetFiber</a></li><li><a href="struct.JanetFunction.html">JanetFunction</a></li><li><a href="struct.JanetGc.html">JanetGc</a></li><li><a href="struct.JanetGcLockGuard.html">JanetGcLockGuard</a></li><li><a href="struct.JanetGcRootGuard.html">JanetGcRootGuard</a></li><li><a href="struct.JanetKeyword.html">JanetKeyword</a></li><li><a href="struct.JanetPointer.html">JanetPointer</a></li><li><a href="struct.JanetRng.html">JanetRng</a></li><li><a href="struct.JanetString.html">JanetString</a></li><li><a href="struct.JanetStruct.html">JanetStruct</a></li><li><a href="struct.JanetSymbol.html">JanetSymbol</a></li><li><a href="struct.JanetTable.html">JanetTable</a></li><li><a href="struct.JanetTuple.html">JanetTuple</a></li><li><a href="allocator/struct.Scratch.html">allocator::Scratch</a></li><li><a href="array/struct.IntoIter.html">array::IntoIter</a></li><li><a href="array/struct.Iter.html">array::Iter</a></li><li><a href="array/struct.IterMut.html">array::IterMut</a></li><li><a href="array/struct.JanetArray.html">array::JanetArray</a></li><li><a href="buffer/struct.JanetBuffer.html">buffer::JanetBuffer</a></li><li><a href="client/struct.JanetClient.html">client::JanetClient</a></li><li><a href="env/struct.CFunOptions.html">env::CFunOptions</a></li><li><a href="env/struct.DefOptions.html">env::DefOptions</a></li><li><a href="env/struct.JanetEnvironment.html">env::JanetEnvironment</a></li><li><a href="env/struct.VarOptions.html">env::VarOptions</a></li><li><a href="fiber/struct.Exec.html">fiber::Exec</a></li><li><a href="fiber/struct.JanetFiber.html">fiber::JanetFiber</a></li><li><a href="function/struct.CallError.html">function::CallError</a></li><li><a href="function/struct.JanetFunction.html">function::JanetFunction</a></li><li><a href="function/struct.JanetTryState.html">function::JanetTryState</a></li><li><a href="io/struct.FileFlags.html">io::FileFlags</a></li><li><a href="io/struct.JanetFile.html">io::JanetFile</a></li><li><a href="janet_abstract/struct.JanetAbstract.html">janet_abstract::JanetAbstract</a></li><li><a href="janet_abstract/struct.JanetAbstractType.html">janet_abstract::JanetAbstractType</a></li><li><a href="pointer/struct.JanetPointer.html">pointer::JanetPointer</a></li><li><a href="string/struct.JanetString.html">string::JanetString</a></li><li><a href="string/struct.JanetStringBuilder.html">string::JanetStringBuilder</a></li><li><a href="structs/struct.IntoIter.html">structs::IntoIter</a></li><li><a href="structs/struct.Iter.html">structs::Iter</a></li><li><a href="structs/struct.JanetStruct.html">structs::JanetStruct</a></li><li><a href="structs/struct.JanetStructBuilder.html">structs::JanetStructBuilder</a></li><li><a href="structs/struct.Keys.html">structs::Keys</a></li><li><a href="structs/struct.Values.html">structs::Values</a></li><li><a href="symbol/struct.JanetKeyword.html">symbol::JanetKeyword</a></li><li><a href="symbol/struct.JanetSymbol.html">symbol::JanetSymbol</a></li><li><a href="table/struct.IntoIter.html">table::IntoIter</a></li><li><a href="table/struct.Iter.html">table::Iter</a></li><li><a href="table/struct.IterMut.html">table::IterMut</a></li><li><a href="table/struct.JanetTable.html">table::JanetTable</a></li><li><a href="table/struct.Keys.html">table::Keys</a></li><li><a href="table/struct.OccupiedEntry.html">table::OccupiedEntry</a></li><li><a href="table/struct.OccupiedError.html">table::OccupiedError</a></li><li><a href="table/struct.VacantEntry.html">table::VacantEntry</a></li><li><a href="table/struct.Values.html">table::Values</a></li><li><a href="table/struct.ValuesMut.html">table::ValuesMut</a></li><li><a href="tuple/struct.IntoIter.html">tuple::IntoIter</a></li><li><a href="tuple/struct.Iter.html">tuple::Iter</a></li><li><a href="tuple/struct.JanetTuple.html">tuple::JanetTuple</a></li><li><a href="tuple/struct.JanetTupleBuilder.html">tuple::JanetTupleBuilder</a></li><li><a href="util/struct.JanetBuildConfig.html">util::JanetBuildConfig</a></li><li><a href="util/struct.JanetVersion.html">util::JanetVersion</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="enum.JanetSignal.html">JanetSignal</a></li><li><a href="enum.JanetType.html">JanetType</a></li><li><a href="enum.TaggedJanet.html">TaggedJanet</a></li><li><a href="client/enum.Error.html">client::Error</a></li><li><a href="fiber/enum.FiberStatus.html">fiber::FiberStatus</a></li><li><a href="function/enum.CallErrorKind.html">function::CallErrorKind</a></li><li><a href="janet_abstract/enum.AbstractError.html">janet_abstract::AbstractError</a></li><li><a href="table/enum.Entry.html">table::Entry</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="trait.DeepEq.html">DeepEq</a></li><li><a href="trait.IsJanetAbstract.html">IsJanetAbstract</a></li><li><a href="trait.JanetArgs.html">JanetArgs</a></li><li><a href="trait.JanetExtend.html">JanetExtend</a></li><li><a href="trait.JanetTypeName.html">JanetTypeName</a></li><li><a href="janet_abstract/trait.IsJanetAbstract.html">janet_abstract::IsJanetAbstract</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.array.html">array</a></li><li><a href="macro.bad_slot.html">bad_slot</a></li><li><a href="macro.check_janet_version.html">check_janet_version</a></li><li><a href="macro.declare_janet_mod.html">declare_janet_mod</a></li><li><a href="macro.janet_mod.html">janet_mod</a></li><li><a href="macro.jcatch.html">jcatch</a></li><li><a href="macro.jpanic.html">jpanic</a></li><li><a href="macro.jtry.html">jtry</a></li><li><a href="macro.structs.html">structs</a></li><li><a href="macro.table.html">table</a></li><li><a href="macro.tuple.html">tuple</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="attr.cjvg.html">cjvg</a></li><li><a href="attr.janet_fn.html">janet_fn</a></li><li><a href="attr.janet_version.html">janet_version</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="janet_abstract/fn.register.html">janet_abstract::register</a></li><li><a href="util/fn.check_fix_arity.html">util::check_fix_arity</a></li><li><a href="util/fn.check_range_arity.html">util::check_range_arity</a></li></ul><h3 id="types">Type Aliases</h3><ul class="all-items"><li><a href="type.JanetCFunction.html">JanetCFunction</a></li><li><a href="array/type.RSplit.html">array::RSplit</a></li><li><a href="array/type.RSplitMut.html">array::RSplitMut</a></li><li><a href="array/type.RSplitN.html">array::RSplitN</a></li><li><a href="array/type.RSplitNMut.html">array::RSplitNMut</a></li><li><a href="array/type.Split.html">array::Split</a></li><li><a href="array/type.SplitMut.html">array::SplitMut</a></li><li><a href="array/type.SplitN.html">array::SplitN</a></li><li><a href="array/type.SplitNMut.html">array::SplitNMut</a></li><li><a href="function/type.JanetCFunction.html">function::JanetCFunction</a></li><li><a href="function/type.JanetRawCFunction.html">function::JanetRawCFunction</a></li><li><a href="tuple/type.RSplit.html">tuple::RSplit</a></li><li><a href="tuple/type.RSplitN.html">tuple::RSplitN</a></li><li><a href="tuple/type.Split.html">tuple::Split</a></li><li><a href="tuple/type.SplitN.html">tuple::SplitN</a></li></ul></section></div></main></body></html>