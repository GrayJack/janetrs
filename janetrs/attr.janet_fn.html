<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Macro that tranforms a high-level Janet function (`fn(&amp;mut [Janet]) -&gt; Janet`) to the thing the Janet C API is expecting (`fn(i32, *mut janetrs::lowlevel::Janet) -&gt; janetrs::lowlevel::Janet`)"><title>janet_fn in janetrs - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-47e7ab555ef2818a.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="janetrs" data-themes="" data-resource-suffix="" data-rustdoc-version="1.74.0-nightly (030e4d382 2023-09-10)" data-channel="nightly" data-search-js="search-5d3eaacf19ebf04f.js" data-settings-js="settings-74424d7eec62a23e.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-f194925aa375ae96.css" data-theme-dark-css="dark-1dd4d1ce031e15de.css" data-theme-ayu-css="ayu-49e58d069f567085.css" ><script src="../static.files/storage-db41da1a38ea3cb8.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-8d035c8cea6edbc4.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-f194925aa375ae96.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1dd4d1ce031e15de.css"><link rel="stylesheet" href="../static.files/noscript-cffde32267a19fd6.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><div class="sidebar-elems"><h2><a href="index.html">In janetrs</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Attribute Macro <a href="index.html">janetrs</a>::<wbr><a class="attr" href="#">janet_fn</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/janetrs_macros/lib.rs.html#46-48">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>#[janet_fn]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Macro that tranforms a high-level Janet function (<code>fn(&amp;mut [Janet]) -&gt; Janet</code>)
to the thing the Janet C API is expecting (<code>fn(i32, *mut janetrs::lowlevel::Janet) -&gt; janetrs::lowlevel::Janet</code>)</p>
<p>The optional argument <code>catch</code> adds code to catch Rust panics and transform them to
Janet panics. This argument is ignored if the <code>std</code> feature is deactivated.</p>
<p>The optional argument <code>arity</code> adds a arity check for the function. It must receive the
kind of arity check. These kinds are <code>fix</code>, for fixed arity, and <code>range</code>, for ranged
or variadic arity. The <code>fix</code> kind receives a integer of the number of the parameters
the Janet function must have and the <code>range</code> kind can receive two arguments, the first
one if mandatory while the second one is optional, the first represents the minimal of
arguments the Janet function have to receive and the second represents the maximum of
arguments the Janet function can receive. If the maximum is not set for the range
arity, the maximum check is disabled, allowing variadic arguments.</p>
<p>The optional arg <code>check_mut_ref</code> adds a check to see if the function received more
than one reference to the same <code>*mut</code> pointer. This check is not the default because
Janet Types act like types with interior mutability and the check is expensive, but if
you want to make sure that your function never receives the same pointer more than
once you can use this.</p>
<p><strong>Usages</strong>:</p>
<ul>
<li><code>#[janet_fn]</code></li>
<li><code>#[janet_fn(arity(fix(&lt;N&gt;)))]</code> where <code>N</code> is an integer literal</li>
<li><code>#[janet_fn(arity(range(&lt;MIN&gt; [, MAX])))]</code> where <code>MIN</code> and <code>MAX</code> are integer
literals</li>
<li><code>#[janet_fn(catch)]</code></li>
<li><code>#[janet_fn(check_mut_ref)]</code></li>
<li><code>#[janet_fn(arity(&lt;...&gt;), check_mut_ref)]</code> Combining both</li>
</ul>
</div></details></section></div></main></body></html>