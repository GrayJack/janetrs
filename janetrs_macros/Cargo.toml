[package]
name = "janetrs_macros"
version = "0.4.6"
authors = ["Eric Shimizu Karbstein <gr41.j4ck@gmail.com>"]
description = "Attribute macros for JanetRS"
repository = "https://github.com/GrayJack/janetrs"
homepage = "https://crates.io/crates/janetrs_macros"
documentation = "https://docs.rs/janetrs_macros"
license-file = "../LICENSE"
edition = "2021"
rust-version = "1.56.0"
keywords = ["janet", "embedding", "proc-macro"]
categories = ["development-tools::procedural-macro-helpers"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
name = "janetrs_macros"
path = "src/lib.rs"
proc-macro = true

[dependencies]
syn = { version = "1", features = ["extra-traits", "full"] }
quote = "1"
proc-macro2 = "1"
janetrs_version = "0.1"

[features]
default = []
system = ["janetrs_version/system"]

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0", features = ["diff"] }
janetrs = { path = "../../janetrs" }
