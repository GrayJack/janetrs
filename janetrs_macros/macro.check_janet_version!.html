<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.check_janet_version.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.check_janet_version.html">macro.check_janet_version.html</a>...</p>
    <script>location.replace("macro.check_janet_version.html" + location.search + location.hash);</script>
</body>
</html>