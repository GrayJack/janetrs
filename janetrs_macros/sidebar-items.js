window.SIDEBAR_ITEMS = {"attr":["cjvg","janet_fn","janet_version"],"macro":["check_janet_version","declare_janet_mod"]};