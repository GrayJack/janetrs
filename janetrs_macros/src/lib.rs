extern crate proc_macro;

use quote::{quote, quote_spanned};
use syn::{parse_macro_input, spanned::Spanned};

/// **Usage**: `#[janet_fn]` or `#[janet_fn(check_mut_ref)]`
///
/// Macro that tranforms a high-level Janet function (`fn(&mut [Janet]) -> Janet`)
/// to the thing the Janet C API is expection (`fn(i32, *mut janetrs::lowlevel::Janet) ->
/// janetrs::lowlevel::Janet`)
///
///
/// The optional arg `check_mut_ref` adds a check to see if the function received more
/// than one reference to the same `*mut` pointer. This check is not the default because
/// Janet Types act like types with interior mutability and the check is expensive, but if
/// you want to make sure that your function never receives the same pointer more than
/// once you can use this.
#[proc_macro_attribute]
pub fn janet_fn(
    args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let func = parse_macro_input!(input as syn::Item);

    let args = parse_macro_input!(args as syn::AttributeArgs);

    if args.len() > 1 {
        return quote! {compile_error!("expected at max one argument to the janet_fn proc-macro");}
            .into();
    }

    let check = if args.is_empty() {
        quote! {}
    } else if let syn::NestedMeta::Meta(syn::Meta::Path(ref path)) = args[0] {
        if let Some(ident) = path.get_ident() {
            if *ident == "check_mut_ref" {
                quote! {
                    for j1 in &args[0..args.len() - 1] {
                        for j2 in &args[1..] {
                            if matches!(
                                j1.kind(),
                                janetrs::types::JanetType::Array
                                    | janetrs::types::JanetType::Buffer
                                    | janetrs::types::JanetType::CFunction
                                    | janetrs::types::JanetType::Fiber
                                    | janetrs::types::JanetType::Function
                                    | janetrs::types::JanetType::Keyword
                                    | janetrs::types::JanetType::Pointer
                                    | janetrs::types::JanetType::Symbol
                                    | janetrs::types::JanetType::Table
                            ) && matches!(
                                j1.kind(),
                                janetrs::types::JanetType::Array
                                    | janetrs::types::JanetType::Buffer
                                    | janetrs::types::JanetType::CFunction
                                    | janetrs::types::JanetType::Fiber
                                    | janetrs::types::JanetType::Function
                                    | janetrs::types::JanetType::Keyword
                                    | janetrs::types::JanetType::Pointer
                                    | janetrs::types::JanetType::Symbol
                                    | janetrs::types::JanetType::Table
                            ) && j1 == j2
                            {
                                janetrs::jpanic!("Received two mutable references as arguments");
                            }
                        }
                    }
                }
            } else {
                return quote_spanned! {ident.span() => compile_error!("Invalid argument. Valid arguments: `check_mut_ref`.");}.into();
            }
        } else {
            quote! {}
        }
    } else {
        quote! {}
    };

    let ts = if let syn::Item::Fn(f) = func {
        // dbg!(&f);
        let f_clone = f.clone();
        let attrs = f.attrs;
        let vis = f.vis;
        let name = f.sig.ident;
        // let syn::ReturnType::Type(_, type_path) = f.sig.output;

        // check for inputs
        if matches!(f.sig.inputs.len(), 1) {
            if let Some(syn::FnArg::Typed(syn::PatType { ty, .. })) = f.sig.inputs.first() {
                if let syn::Type::Reference(syn::TypeReference {
                    mutability,
                    ref elem,
                    ..
                }) = **ty
                {
                    if mutability.is_none() {
                        return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found a immutable reference");}.into();
                    } else if let syn::Type::Slice(syn::TypeSlice {
                        elem: ref slice, ..
                    }) = **elem
                    {
                        if let syn::Type::Path(syn::TypePath { ref path, .. }) = **slice {
                            if !janet_path_checker(path) {
                                return quote_spanned! {path.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                            }
                        } else {
                            return quote_spanned! {slice.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                        }
                    } else {
                        return quote_spanned! {elem.span() => compile_error!("expected to be a slice of `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found something that is not a reference at all");}.into();
                }
            }

            // check output type
            let output_span = match f.sig.output.clone() {
                syn::ReturnType::Default => f_clone.sig.span(),
                syn::ReturnType::Type(_, ty) => ty.span(),
            };
            if let syn::ReturnType::Type(_, ty) = f.sig.output {
                if let syn::Type::Path(syn::TypePath { ref path, .. }) = *ty {
                    if !janet_path_checker(path) {
                        return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                }
            } else {
                return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
            }

            quote! {
                #(#attrs)*
                #[no_mangle]
                #vis unsafe extern "C" fn #name(argc: i32, argv: *mut ::janetrs::lowlevel::Janet) -> ::janetrs::lowlevel::Janet {
                    #[inline]
                    #f_clone

                    let args = unsafe { core::slice::from_raw_parts_mut(argv, argc as usize) };
                    let mut args = unsafe { &mut *(args as *mut [::janetrs::lowlevel::Janet] as *mut [::janetrs::types::Janet])};

                    #check

                    #name(args).into()
                }
            }
        } else {
            quote_spanned! {f_clone.sig.inputs.span() => compile_error!("expected exactly one argument of type `&mut [janetrs::types::Janet]`");}
        }
    } else {
        quote_spanned! {func.span() => compile_error!("expected fn item");}
    };

    ts.into()
}

fn janet_path_checker(path: &syn::Path) -> bool {
    match path.segments.len() {
        1 => {
            let ident = if let Some(i) = path.get_ident() {
                i
            } else {
                return false;
            };
            let test = syn::Ident::new("Janet", ident.span());

            *ident == test
        },
        2 => {
            let mod_types = &path.segments.first().unwrap().ident;
            let janet_ident = &path.segments.last().unwrap().ident;

            let mod_expected = syn::Ident::new("types", mod_types.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *mod_types == mod_expected && *janet_ident == janet_expected
        },
        3 => {
            let mut iter = path.segments.iter();
            let (janetrs_mod, types_mod, janet_ident) = (
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
            );

            let janetrs_expected = syn::Ident::new("janetrs", janetrs_mod.span());
            let types_expected = syn::Ident::new("types", types_mod.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *janetrs_mod == janetrs_expected
                && *types_mod == types_expected
                && *janet_ident == janet_expected
        },
        _ => false,
    }
}
