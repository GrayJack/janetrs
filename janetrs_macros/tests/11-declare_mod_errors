use janetrs_macros::*;
use janetrs::{self, Janet};

#[janet_fn]
pub fn function(_args: &mut [Janet]) -> Janet {
    Janet::nil()
}

#[janet_fn]
pub fn function2(_args: &mut [janetrs::Janet]) -> janetrs::Janet {
    Janet::nil()
}

declare_janet_mod!(10;
    {"function1", function},
    {"function2", function2},
);

declare_janet_mod!("test";
    {10, function},
    {"function2", function2},
);

declare_janet_mod!("test";
    {"function1", 10},
    {"function2", function2},
);

declare_janet_mod!("test";
    {"function1", function},
    {"function2", "function2"},
);

fn main() {}