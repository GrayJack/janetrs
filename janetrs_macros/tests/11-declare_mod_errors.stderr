error: expected string literal
  --> $DIR/11-declare_mod_errors:14:20
   |
14 | declare_janet_mod!(10;
   |                    ^^

error: expected string literal
  --> $DIR/11-declare_mod_errors:20:6
   |
20 |     {10, function},
   |      ^^

error: expected identifier
  --> $DIR/11-declare_mod_errors:25:19
   |
25 |     {"function1", 10},
   |                   ^^

error: expected identifier
  --> $DIR/11-declare_mod_errors:31:19
   |
31 |     {"function2", "function2"},
   |                   ^^^^^^^^^^^
